release/src/Upload.hs

```diff
@@ -0,0 +1,71 @@
+-- Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE MultiWayIf #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE StrictData #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Upload (
+    uploadToMavenCentral
+) where
+
+import qualified Control.Exception.Safe as E
+import           Control.Monad.Logger
+import           Control.Monad.IO.Class
+import           Data.Foldable (for_)
+import           Data.List (foldl)
+import           Network.HTTP.Client (newManager, httpNoBody)
+import           Network.HTTP.Client.TLS (tlsManagerSettings)
+import           Network.HTTP.Simple (getResponseStatusCode, parseRequest, setRequestBasicAuth, setRequestBodyFile, setRequestHeader, setRequestMethod, setRequestPath)
+import           Data.Text (Text, pack)
+import           Data.Text.Encoding (encodeUtf8)
+import           Path
+import           Path.IO
+
+import Types
+import Util (BazelLocations(..))
+
+uploadToMavenCentral ::(MonadLogger m, MonadIO m, E.MonadThrow m) => PerformUpload -> BazelLocations -> [(MavenCoords, Path Rel File)] -> m ()
+uploadToMavenCentral upload BazelLocations{..} artifacts = do
+
+    manager <- liftIO $ newManager tlsManagerSettings
+
+    initialRequest' <- parseRequest "http://localhost:8081"
+
+    for_ artifacts $ \(coords@MavenCoords{..}, file) -> do
+        binExists <- doesFileExist (bazelBin </> file)
```

Comment:
It looks like you are uploading the files directly from the Bazel output directory instead of uploading them from the release directory like we did for Bintray. Any reason for that? I’d prefer the latter as it puts the files in the right locations and you will get an error even in dry run if the file to be uploaded does not exist.

Reply:
It's more or less just a designed decision as I wanted the Maven Coordinates available for uploading, then not having the files in the right locations isn't a problem.

I could either
a) read from release directory (assuming that everything needed will always be there) o
or
b) check the file existence (and possible for read permissions) when in dry run mode.

WDYT?

Reply:
I would go with a) given that this was who we handled things for the bintray upload.

Reply:
Implemented.